src/os_cloud_nodelet.cpp

```diff
@@ -72,53 +73,74 @@ class OusterCloud : public nodelet::Nodelet {
             tf_bcast.broadcast_transforms(info);
         } else {
             NODELET_INFO_STREAM(
-                "OusterCloud: dynamic transforms broadcast enabled wit "
+                "OusterCloud: dynamic transforms broadcast enabled with "
                 "broadcast rate of: "
                 << dynamic_transforms_rate << " Hz");
+            timer_.stop();
             timer_ = getNodeHandle().createTimer(
                 ros::Duration(1.0 / dynamic_transforms_rate),
                 [this, info](const ros::TimerEvent&) {
                     tf_bcast.broadcast_transforms(info, last_msg_ts);
                 });
         }
 
-        create_publishers_subscribers(info);
+        create_handlers(info);
+    }
+
+    void create_imu_pub_sub() {
+        imu_pub = getNodeHandle().advertise<sensor_msgs::Imu>("imu", 100);
+        imu_packet_sub = getNodeHandle().subscribe<PacketMsg>(
+            "imu_packets", 100, [this](const PacketMsg::ConstPtr msg) {
+                if (imu_packet_handler) {
+                    auto imu_msg = imu_packet_handler(msg->buf.data());
+                    if (imu_msg.header.stamp > last_msg_ts)
+                        last_msg_ts = imu_msg.header.stamp;
+                    imu_pub.publish(imu_msg);
+                }
+            });
     }
 
-    void create_publishers_subscribers(const sensor::sensor_info& info) {
+    void create_point_cloud_pubs() {
+        // NOTE: always create the 2nd topic
+        lidar_pubs.resize(2);
+        for (int i = 0; i < 2; ++i) {
+            lidar_pubs[i] = getNodeHandle().advertise<sensor_msgs::PointCloud2>(
+                topic_for_return("points", i), 10);
+        }
+    }
+
+    void create_laser_scan_pubs() {
+        // NOTE: always create the 2nd topic
+        scan_pubs.resize(2);
+        for (int i = 0; i < 2; ++i) {
+            scan_pubs[i] = getNodeHandle().advertise<sensor_msgs::LaserScan>(
+                topic_for_return("scan", i), 10);
+        }
+    }
+
+    void create_lidar_packets_sub() {
+        lidar_packet_sub = getNodeHandle().subscribe<PacketMsg>(
+        "lidar_packets", 100, [this](const PacketMsg::ConstPtr msg) {
+            if (lidar_packet_handler) lidar_packet_handler(msg->buf.data());
+        });
+    }
+
+    void create_handlers(const sensor::sensor_info& info) {
         auto& pnh = getPrivateNodeHandle();
         auto proc_mask = pnh.param("proc_mask", std::string{"IMU|PCL|SCAN"});
         auto tokens = impl::parse_tokens(proc_mask, '|');
 
         auto timestamp_mode = pnh.param("timestamp_mode", std::string{});
         double ptp_utc_tai_offset = pnh.param("ptp_utc_tai_offset", -37.0);
 
-        auto& nh = getNodeHandle();
-
         if (impl::check_token(tokens, "IMU")) {
-            imu_pub = nh.advertise<sensor_msgs::Imu>("imu", 100);
```

Comment:
It seems like you are always creating the publishers now. Do you mean to do that?

If so we should get rid of the proc_mask and just detect if one of the topics are subscribed to before generating/publishing it.

Reply:
I actually meant to create the topics only if their respective token was set. But forgot to add a check before creating the respective topic for each processor.

Reply:
addressed